File: Java/Dagger/atm/src/main/java/ro/razvan/java/dagger/atm/DepositCommand.java
package ro.razvan.java.dagger.atm;

import javax.inject.Inject;
import java.math.BigDecimal;

public final class DepositCommand implements BigDecimalCommand {

    public static final String COMMAND_KEY_DEPOSIT = "deposit";

    private final Outputter outputter;
    private final Database.Account account;
    private final WithdrawalLimiter withdrawalLimiter;

    @Inject
    public DepositCommand(
            Outputter outputter,
            Database.Account account,
            WithdrawalLimiter withdrawalLimiter
    ) {
        this.outputter = outputter;
        this.account = account;
        this.withdrawalLimiter = withdrawalLimiter;
    }

    @Override
    public String key() {
        return COMMAND_KEY_DEPOSIT;
    }

    @Override
    public void handleAmount(BigDecimal amount) {
        account.deposit(amount);
        withdrawalLimiter.recordDeposit(amount);
        outputter.output(account.username() + " now has: " + account.balance());
    }
}

File: Kotlin/Dagger/atm/src/main/kotlin/ro/razvan/kotlin/dagger/atm/DepositCommand.kt
package ro.razvan.kotlin.dagger.atm

import java.math.BigDecimal
import javax.inject.Inject

const val COMMAND_KEY_DEPOSIT = "deposit"

class DepositCommand @Inject constructor(
    private val outputter: Outputter,
    private val account: Database.Account
) : BigDecimalCommand {

    override fun handleAmount(amount: BigDecimal) {
        account.deposit(amount)
        outputter("${account.username} now has: ${account.balance}")
    }

    override val key: String
        get() = COMMAND_KEY_DEPOSIT

}